### PR TITLE
Report Rails server startup failures

### PR DESCRIPTION
**Pull request by justin808**
## Summary

Rails server startup failures now surface as `CypressOnRails::ServerError` with the exact command, process status, and a bounded tail of recent output. Direct spawn errors, immediate exits, and readiness timeouts retain useful diagnostics while cleanup remains bounded and safe around process-reaping races, stalled terminals, broken pipes, and incompatible terminal encodings.

This is delivery 1 only. Shutdown configuration, port selection, Puma behavior, state reset, and #186 remain out of scope.

Related: #185

## Validation

- `bundle exec rspec spec/cypress_on_rails/server_spec.rb` — 28 examples, 0 failures
- `bundle exec rake` — 99 examples, 0 failures
- Ruby 2.7.8 and current-Ruby syntax checks — passed for both changed files
- Independent Sol/xhigh closeout review — CLEAN after lifecycle, timeout, output, and signal-race regressions
- `git diff --check` — passed

### Interrupt-safe process reaping

Reaping a child and caching its exit status must behave as one indivisible transition. The readiness timeout therefore raises an explicit `Timeout::Error`, while the short `waitpid2`/cache section defers that exact exception:

```ruby
# Preserve the status before delivering a pending readiness timeout.
Thread.handle_interrupt(Timeout::Error => :never) do
  result = Process.waitpid2(pid, Process::WNOHANG)
  cache_status(result.last) if result
end
```

This avoids signaling a PID after its child was already reaped and keeps failure messages accurate at timeout boundaries. It is appropriate for small, bounded state transitions; it should not wrap blocking cleanup that could hang indefinitely.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Codex

## Summary by CodeRabbit

* **New Features**
  * Added a dedicated server startup/shutdown error type with richer context.
  * Bounded and more resilient capture/forwarding of server startup output (bytes/lines) with improved handling for terminal and pipe failures.
* **Bug Fixes**
  * Improved server startup readiness handling, including clearer outcomes for early exits vs timeouts.
  * More reliable process-group-aware shutdown with safe escalation when needed.
  * Prevented leaked processes, pipes, threads, and ensured captured diagnostics are preserved.
* **Tests**
  * Added an extensive RSpec test suite covering lifecycle edge cases and output-capture/forwarding behavior.

## Agent Merge Confidence

- Current head: `928d5b9d826c6f8df90b4b51d92e84c1f8fd0ac9`
- Local validation: focused server spec 28 examples/0 failures; full `bundle exec rake` 99 examples/0 failures; Ruby 2.7/current syntax and diff checks passed.
- Independent review: Sol/xhigh closeout review CLEAN after real process-group lifecycle reproductions.
- Hosted review: current-head Claude and CodeRabbit checks passed; all review threads resolved.
- Batch QA: PASS with runtime, workflow/docs integration, clean-apply, and merge-tree evidence; no blockers, non-blockers, or UNKNOWN facts.
- Current-head CI: READY; Rails 6.1, Rails 7.2, Rails 8, Claude review, and CodeRabbit green. Rails 6.1 retry passed after the first attempt was canceled during externally slow Ubuntu mirror downloads.
- Merge ledger: base `07bf729f5a29ff71fcedc625186ac3d2da95b349`; exactly `lib/cypress_on_rails/server.rb` and `spec/cypress_on_rails/server_spec.rb`; unresolved review threads 0; pending/failing/unknown gates 0; complete allowed.
- Release mode: issue #185 delivery 1 only; no CHANGELOG change because the batch file map excludes it.
- Merge authority: `auto_merge_when_gates_pass`.

### CODE REVIEW

**Comment by justin808** [comment]
# QA Evidence

```qa-evidence v1
batch: cpor-0710-docs-spawn
qa_lane: cpor-qa-arc
worktree: /Users/justin/.codex/worktrees/cpor-0710-docs-spawn/qa
base_origin_master: d5566c85933f7b0dd1bc439707ab9b2fcd1b9a44
docs_pr_242_head: 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282
spawn_pr_243_head: 3d6f807fbad534cf59e9d245d1b24ab41b86b9d5
timestamps_utc:
  docs_checkout_start: 2026-07-13T10:04:47Z
  docs_rake: 2026-07-13T10:05:05Z..2026-07-13T10:05:07Z
  spawn_checkout_start: 2026-07-13T10:05:15Z
  spawn_rake: 2026-07-13T10:05:41Z..2026-07-13T10:05:47Z
  combined_check: 2026-07-13T10:06:05Z

exact_sha_reestablishment:
  docs: "git checkout --detach 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282; git rev-parse HEAD -> 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282 (exit 0)"
  spawn: "git checkout --detach 3d6f807fbad534cf59e9d245d1b24ab41b86b9d5; git rev-parse HEAD -> 3d6f807fbad534cf59e9d245d1b24ab41b86b9d5 (exit 0)"

docs_scope_and_commands:
  scope: "git diff --name-only base..docs -> .github/workflows/trigger-docs-site.yml; script/check-docs-sidebar only (exit 0)"
  diff_check: "git diff --check base..docs -> clean (exit 0)"
  invalid_base: "script/check-docs-sidebar definitely-not-a-ref HEAD -> exit 1; Error: cannot resolve base ref: definitely-not-a-ref"
  invalid_current: "script/check-docs-sidebar base definitely-not-a-ref -> exit 1; Error: cannot resolve current ref: definitely-not-a-ref"
  unrelated_history: "script/check-docs-sidebar 7dc7581de07267f4307a5719262fbb64db340f4c HEAD -> exit 1; Error: cannot find merge base between … and HEAD"
  valid_comparison: "script/check-docs-sidebar base HEAD -> exit 0; OK No new published doc files detected"
  bash_syntax: "bash -n script/check-docs-sidebar -> exit 0"
  shellcheck: "shellcheck script/check-docs-sidebar -> installed; exit 0"
  actionlint: "actionlint .github/workflows/trigger-docs-site.yml -> installed; exit 0"
  yaml_parse: "ruby -e YAML.load_file(workflow) -> yaml=ok; exit 0"
  rake: "bundle exec rake -> exit 0; 71 examples, 0 failures"
  workflow_audit: "push on master limited to docs/** and README.md; permissions contents: read; secret references limited to DOCS_DISPATCH_APP_ID and DOCS_DISPATCH_APP_KEY; actions/create-github-app-token pinned d72941d797fd3113feb6b93fd0dec494b13a2547 and peter-evans/repository-dispatch pinned ff45666b9427631e3450c54a1bcbee4d9ff4d7c0; no dispatch was executed."
  diagnostics_audit: "script independently rejects unresolved base ref, unresolved current ref, missing merge-base, and diff-comparison failure with distinct fail-closed errors."

spawn_scope_and_commands:
  scope: "git diff --name-only base..spawn -> lib/cypress_on_rails/server.rb; spec/cypress_on_rails/server_spec.rb only (exit 0)"
  focused_spec: "bundle exec rspec spec/cypress_on_rails/server_spec.rb -> exit 0; 24 examples, 0 failures"
  rake: "bundle exec rake -> exit 0; 95 examples, 0 failures"
  ruby_27_syntax: "mise exec ruby@2.7.8 -- ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  current_ruby_syntax: "ruby 3.4.6; ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  diff_check: "git diff --check base..spawn -> clean (exit 0)"
  runtime_observations: "Focused passing examples explicitly cover direct spawn error/resource cleanup; immediate exit command/status/output; readiness timeout and cached real-timeout status; ECHILD; forwarding EPIPE and EncodingError; bounded pathological lines and multibyte output; lagging and stalled terminal output cleanup; pre-KILL ECHILD/no-KILL and live/exactly-one-KILL; timeout-boundary zombie status; and reader/forwarder thread cleanup."
  out_of_scope_behavior_audit: "Reviewed changed paths/diff; no shutdown, port-management, Puma, state-reset, or issue-186 behavior was added."

combined_interaction:
  path_intersection: "comm of sorted changed-path sets -> empty; docs and spawn paths are disjoint"
  merge_tree: "git merge-tree base docs spawn -> exit 0, clean merge result; no conflict records"
  worktree_status: "git status --short --branch -> ## HEAD (no branch); no repository files modified"

findings:
  blockers: []
  non_blockers: []
  unknowns: []
verdict: PASS
```

File: lib/cypress_on_rails/server.rb [hunk 1]
```diff
@@ -151,22 +187,254 @@ def server_responding?
...
+      ServerError.new(message)
+    end
+
+    def server_start_timeout_failure(timeout)
+      if server_exited?
+        status = startup_process_status
+        drain_server_output
+      else
+        status = process_exists?(@server_pid) ? 'process status running' : 'process status unavailable'
+        stop_result = stop_server(@server_pid)
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/cypress_on_rails/server.rb]: Fixed at 67c0db4. Readiness timeouts now flow through start_server ensure before ServerError construction, so before_server_stop runs before TERM/KILL and cleanup is performed exactly once. The focused regression asserts hook-before-TERM ordering while retaining command, process status, and final bounded output; the full server spec is 28 examples with 0 failures.

**Comment by justin808** [comment]
Current-head review response (67c0db4):

- Preserved the process-group identity directly from the `pgroup: true` spawn PID, including fast-launcher exit paths.
- Cleanup now uses negative-PGID TERM and bounded group-existence polling, escalating to negative-PGID KILL for TERM-resistant descendants whether the leader was already terminal or exits under TERM.
- Readiness-timeout failure construction is deferred until `start_server` cleanup completes, preserving `before_server_stop` before signaling and avoiding duplicate cleanup.
- Added focused regressions for fast-launcher groups, terminal-leader escalation, live-to-terminal escalation, and hook-before-signal ordering.

Validation: focused server spec 28 examples / 0 failures; `bundle exec rake` 99 examples / 0 failures; Ruby 2.7.8 and current Ruby syntax checks pass; independent Sol/xhigh review is CLEAN; strict security preflight passes. A CHANGELOG edit is intentionally excluded by the batch exact-path map and no trusted repository-local policy was found requiring one for this partial delivery.

**Comment by justin808** [comment]
# QA Evidence

```qa-evidence v1
batch: cpor-0710-docs-spawn
qa_lane: cpor-qa-arc
qa_worktree: /Users/justin/.codex/worktrees/cpor-0710-docs-spawn/qa
validation_clone: /tmp/cpor-qa-current-head (fresh independent local clone; writable metadata)
base_origin_master: d5566c85933f7b0dd1bc439707ab9b2fcd1b9a44
docs_pr_242_head: 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282
spawn_pr_243_head: 67c0db471339b32a54fe5af97d17416eb1aa79c9
utc:
  docs_detached_checkout: 2026-07-13T10:40:43Z
  spawn_detached_checkout: 2026-07-13T10:41:12Z
  combined_check: 2026-07-13T10:41:42Z

exact_sha_reestablishment:
  clone: "git clone --no-local --no-checkout /Users/justin/codex/cypress-playwright-on-rails /tmp/cpor-qa-current-head -> exit 0"
  docs: "git checkout --detach 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282; git rev-parse HEAD -> 4e4657a42bd0bb43bc6bd2aa5fc31bf52c3d3282 (exit 0)"
  spawn: "git checkout --detach 67c0db471339b32a54fe5af97d17416eb1aa79c9; git rev-parse HEAD -> 67c0db471339b32a54fe5af97d17416eb1aa79c9 (exit 0)"

docs_scope_commands_and_audit:
  scope: "git diff --name-only base..docs -> .github/workflows/trigger-docs-site.yml; script/check-docs-sidebar only (exit 0)"
  diff_check: "git diff --check base..docs -> clean (exit 0)"
  invalid_base: "script/check-docs-sidebar definitely-not-a-ref HEAD -> exit 1; Error: cannot resolve base ref"
  invalid_current: "script/check-docs-sidebar base definitely-not-a-ref -> exit 1; Error: cannot resolve current ref"
  unrelated_history: "script/check-docs-sidebar <ephemeral unreferenced clone commit 09a3d42b1d4fe3638d376d1aa5c632d02517f205> HEAD -> exit 1; Error: cannot find merge base"
  valid_comparison: "script/check-docs-sidebar base HEAD -> exit 0; OK No new published doc files detected"
  bash_syntax: "bash -n script/check-docs-sidebar -> exit 0"
  shellcheck: "shellcheck script/check-docs-sidebar -> exit 0"
  actionlint: "actionlint .github/workflows/trigger-docs-site.yml -> exit 0"
  yaml_parse: "ruby -e YAML.load_file(workflow) -> YAML OK; exit 0"
  rake: "bundle exec rake -> exit 0; 71 examples, 0 failures"
  workflow: "push trigger is master only and paths docs/** plus README.md; permissions are contents: read; secret references are DOCS_DISPATCH_APP_ID and DOCS_DISPATCH_APP_KEY only; actions/create-github-app-token is SHA-pinned d72941d797fd3113feb6b93fd0dec494b13a2547 and peter-evans/repository-dispatch is SHA-pinned ff45666b9427631e3450c54a1bcbee4d9ff4d7c0. No secret-bearing dispatch was executed."
  fail_closed_diagnostics: "Inspected script errors: distinct failures for unresolved base ref, unresolved current ref, missing merge base, and failed comparison diff."

spawn_scope_commands_and_runtime:
  scope: "git diff --name-only base..spawn -> lib/cypress_on_rails/server.rb; spec/cypress_on_rails/server_spec.rb only (exit 0)"
  diff_check: "git diff --check base..spawn -> clean (exit 0)"
  ruby_27_syntax: "mise exec ruby@2.7.8 -- ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  current_ruby_syntax: "ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  focused_spec: "bundle exec rspec spec/cypress_on_rails/server_spec.rb -> exit 0; 28 examples, 0 failures"
  rake: "bundle exec rake -> exit 0; 99 examples, 0 failures"
  runtime_coverage: "Passing focused suite covers direct spawn and partial-capture errors; immediate exit status/output; readiness and live timeout status caching; ECHILD; pre-KILL no-KILL and live exactly-one-KILL; timeout-boundary status; EPIPE/EncodingError forwarding; bounded lines/bytes and multibyte output; lagging/stalled/draining output; reader/forwarder cleanup; fast-launcher PGID retention; terminal-leader TERM-resistant group escalation; live-to-terminal TERM-resistant group escalation; and before_server_stop before signal ordering."
  scope_expansion_audit: "Changed-diff scan found no shutdown policy, port-management, Puma, state-reset, issue-186, docs, or CHANGELOG scope expansion."

combined_interaction:
  path_disjoint: "comm of sorted docs and spawn changed-path sets -> empty"
  merge_tree: "git merge-tree base docs spawn -> exit 0; no conflict markers or CONFLICT records"
  status: "independent clone and QA worktree both report ## HEAD (no branch) with no modified repository files"

findings:
  blockers: []
  non_blockers: []
  unknowns: []
verdict: PASS
```

File: lib/cypress_on_rails/server.rb [hunk 2]
```diff
@@ -151,33 +189,315 @@ def server_responding?
...
+      loop do
+        queue.push(output, true)
+        return [true, deadline]
+      rescue ThreadError
+        deadline ||= monotonic_time + SERVER_OUTPUT_FORWARD_BACKPRESSURE_TIMEOUT
+        return false if monotonic_time >= deadline
+
+        sleep 0.001
+      end
+    end
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/cypress_on_rails/server.rb:445]: Triaged as non-blocking maintainability feedback. Ruby multiple assignment intentionally gives forwarding=false and forwarding_deadline=nil here; once forwarding is false the deadline is not consumed again. Current regressions cover the bounded backpressure and forwarding shutdown behavior, so this partial delivery keeps the reviewed current head unchanged.

File: lib/cypress_on_rails/server.rb [hunk 3]
```diff
@@ -2,10 +2,23 @@
 require 'timeout'
 require 'fileutils'
 require 'net/http'
+require 'thread'
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/cypress_on_rails/server.rb:5]: Triaged as non-blocking cleanup. Keeping the explicit standard-library require is harmless, documents the synchronization dependency, and preserves compatibility clarity across the supported Ruby range. No behavior change is warranted in this delivery.

File: lib/cypress_on_rails/server.rb [hunk 4]
```diff
@@ -151,33 +189,315 @@ def server_responding?
...
+          unless server_exited?
+            send_kill_signal(pid)
+            wait_for_server_exit(monotonic_time + SERVER_STOP_TIMEOUT)
+          end
+        end
+        :signaled
+      ensure
+        wait_for_server_output
+      end
+    end
```
[the remark this answers is not in the document]
Comment (justin808) [reply lib/cypress_on_rails/server.rb:219]: Confirmed intentional defensive coverage. Successful pgroup spawn establishes the normal PGID path; the nil-PGID branches preserve safe behavior for partial/direct lifecycle states and focused test seams without expanding shutdown policy. Current runtime coverage exercises both group and fallback safety paths.

**Comment by justin808** [comment]
# QA Evidence

```qa-evidence v1
batch: cpor-0710-docs-spawn
qa_lane: cpor-qa-arc
qa_worktree: /Users/justin/.codex/worktrees/cpor-0710-docs-spawn/qa
validation_clone: /tmp/cpor-qa-final-post-docs (fresh independent local clone with writable metadata)
baseline_before_docs_correction: d5566c85933f7b0dd1bc439707ab9b2fcd1b9a44
merged_docs_current_master: 07bf729f5a29ff71fcedc625186ac3d2da95b349
spawn_current_head: 928d5b9d826c6f8df90b4b51d92e84c1f8fd0ac9
utc:
  docs_detached_checkout: 2026-07-13T10:48:22Z
  spawn_detached_checkout: 2026-07-13T10:48:46Z
  final_topology_and_apply_check: 2026-07-13T10:49:36Z

exact_sha_reestablishment:
  clone: "git clone --no-local --no-checkout /Users/justin/codex/cypress-playwright-on-rails /tmp/cpor-qa-final-post-docs -> exit 0"
  docs: "git checkout --detach 07bf729f5a29ff71fcedc625186ac3d2da95b349; git rev-parse HEAD -> 07bf729f5a29ff71fcedc625186ac3d2da95b349 (exit 0)"
  spawn: "git checkout --detach 928d5b9d826c6f8df90b4b51d92e84c1f8fd0ac9; git rev-parse HEAD -> 928d5b9d826c6f8df90b4b51d92e84c1f8fd0ac9 (exit 0)"

merged_docs_checks:
  scope: "git diff --name-only baseline..merged-docs -> .github/workflows/trigger-docs-site.yml; script/check-docs-sidebar only (exit 0)"
  reviewed_content_match: "git diff --exit-code original-reviewed-docs..merged-docs -- two mapped paths -> exit 0; contents match"
  diff_check: "git diff --check baseline..merged-docs -> clean (exit 0)"
  invalid_base: "script/check-docs-sidebar definitely-not-a-ref HEAD -> exit 1; cannot resolve base ref"
  invalid_current: "script/check-docs-sidebar baseline definitely-not-a-ref -> exit 1; cannot resolve current ref"
  unrelated_history: "script/check-docs-sidebar ephemeral-unreferenced-clone-commit HEAD -> exit 1; cannot find merge base"
  valid_comparison: "script/check-docs-sidebar baseline HEAD -> exit 0; OK No new published doc files detected"
  bash_syntax: "bash -n script/check-docs-sidebar -> exit 0"
  shellcheck: "shellcheck script/check-docs-sidebar -> exit 0"
  actionlint: "actionlint .github/workflows/trigger-docs-site.yml -> exit 0"
  yaml_parse: "ruby -e YAML.load_file(workflow) -> YAML OK; exit 0"
  rake: "bundle exec rake -> exit 0; 71 examples, 0 failures"
  workflow_audit: "trigger is push to master restricted to docs/** and README.md; permissions are contents: read; secret references are DOCS_DISPATCH_APP_ID and DOCS_DISPATCH_APP_KEY only; action SHAs are actions/create-github-app-token@d72941d797fd3113feb6b93fd0dec494b13a2547 and peter-evans/repository-dispatch@ff45666b9427631e3450c54a1bcbee4d9ff4d7c0. No dispatch was executed."
  diagnostics_audit: "distinct fail-closed script diagnostics exist for base-ref, current-ref, merge-base, and comparison-diff failures."

spawn_current_head_checks:
  merge_base: "git merge-base merged-docs HEAD -> 07bf729f5a29ff71fcedc625186ac3d2da95b349"
  docs_ancestor: "git merge-base --is-ancestor merged-docs HEAD -> exit 0"
  pr_scope: "git diff --name-only merged-docs..spawn -> lib/cypress_on_rails/server.rb; spec/cypress_on_rails/server_spec.rb only (exit 0)"
  docs_delta: "git diff --name-only merged-docs..spawn -- docs mapped paths -> empty"
  diff_check: "git diff --check merged-docs..spawn -> clean (exit 0)"
  ruby_27_syntax: "mise exec ruby@2.7.8 -- ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  current_ruby_syntax: "ruby -c lib/cypress_on_rails/server.rb -> Syntax OK; exit 0"
  focused_spec: "bundle exec rspec spec/cypress_on_rails/server_spec.rb -> exit 0; 28 examples, 0 failures"
  rake: "bundle exec rake -> exit 0; 99 examples, 0 failures"
  runtime_coverage: "Passing focused suite covers direct spawn/partial-capture errors; immediate exit status/output; readiness/live timeout status caching; ECHILD; pre-KILL no-KILL and live exactly-one-KILL; timeout-boundary status; EPIPE/EncodingError forwarding; bounded lines/bytes and multibyte output; lagging/stalled/draining output; reader/forwarder cleanup; fast-launcher PGID retention; terminal-leader TERM-resistant group escalation; live-to-terminal TERM-resistant group escalation; and before_server_stop before signal ordering."
  scope_audit: "Changed-diff scan found no shutdown policy, port-management, Puma, state-reset, issue-186, docs, or CHANGELOG scope expansion."

current_topology_and_apply:
  apply_check: "from detached merged-docs, git diff --no-ext-diff --binary merged-docs..spawn | git apply --check -> exit 0"
  merge_tree: "git merge-tree merged-docs merged-docs spawn -> exit 0; no conflict records or markers"
  statuses: "fresh clone and QA worktree both report ## HEAD (no branch), with no modified repository files"

findings:
  blockers: []
  non_blockers: []
  unknowns: []
verdict: PASS
```

**Comment by justin808** [comment]
QA replay marker for batch closeout. The detailed human-readable evidence remains in the prior QA comment.
